patches/tModLoader/Terraria/Localization/LanguageManager.tML.cs

```diff
@@ -148,4 +150,75 @@ private string ComputeBoundTextValue(TextBinding binding)
 		return string.Format(value, binding.args);
 	}
 	#endregion
+
+	private const int defaultWatcherCooldown = 15;
+	private Dictionary<Mod, FileSystemWatcher> localizationFileWatchers = new();
+	internal HashSet<string> changedFiles = new();
+	internal HashSet<string> pendingFiles = new();
+	internal int watcherCooldown;
+	internal void SetupFileWatchers()
+	{
+		// Add a watcher for each loaded mod that has a corresponding mod sources folder
+		// Don't worry about the mod being local or not, for now. The feature might be useful for even workshop tmod files
+		foreach (var mod in ModLoader.ModLoader.Mods) {
+			string path = Path.Combine(ModLoader.Core.ModCompile.ModSourcePath, mod.Name);
+			if (!Directory.Exists(path))
+				continue;
+
+			try {
+				var localizationFileWatcher = new FileSystemWatcher();
+				localizationFileWatcher.Path = path;
+				localizationFileWatcher.NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.FileName;
+				localizationFileWatcher.Filter = "*.hjson";
+				localizationFileWatcher.IncludeSubdirectories = true;
+
+				localizationFileWatcher.Changed += (a, b) => {
+					HandleFileChangedOrRenamed(mod.Name, b.Name);
+				};
+				localizationFileWatcher.Renamed += (a, b) => {
+					HandleFileChangedOrRenamed(mod.Name, b.Name);
+				};
+
+				// Begin watching.
+				localizationFileWatcher.EnableRaisingEvents = true;
+
+				localizationFileWatchers[mod] = localizationFileWatcher;
+			}
+			catch (Exception) {
+				throw;
+			}
+		}
+	}
+
+	internal void UnloadFileWatchers()
+	{
+		foreach (var fileWatcher in localizationFileWatchers.Values) {
+			fileWatcher.EnableRaisingEvents = false;
+			fileWatcher.Dispose();
+		}
+		localizationFileWatchers.Clear();
+	}
+
+	private void HandleFileChangedOrRenamed(string modName, string fileName)
+	{
+		watcherCooldown = defaultWatcherCooldown;
+		pendingFiles.Add(Path.Combine(modName, fileName));
```

Comment:
I think you need to `lock` around accesses to this collection. If you breakpoint you'll find that this method and `Update` are called on different threads.